[NVPTX] Improve 64bit FSH/ROT lowering when shift amount is constant

AlexMaclean opened the pull request:
When the sift amount of a 64-bit funnel-shift or rotate is constant, it may be decomposed into two 32-bit funnel-sifts. This ensures that we recover any possible performance losses associated with the correctness fix in a131fbf1.

In order to efficiently represent the expansion with Selection DAG nodes, NVPTXISD::BUILD_VECTOR and NVPTXISD::UNPACK_VECTOR are added which allow the vector output/input to be represented as a scalar. In the future, if we add support for the v2i32 type to the NVPTX backend these nodes may be removed.

AlexMaclean committed:
[NVPTX] Improve 64-bit FSH/ROT lowering when the shift ammount is a constant

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -445,6 +448,17 @@ bool NVPTXDAGToDAGISel::SelectSETP_BF16X2(SDNode *N) {
   return true;
 }
 
+bool NVPTXDAGToDAGISel::tryUNPACK_VECTOR(SDNode *N) {
+  SDValue Vector = N->getOperand(0);
+  MVT EltVT = N->getSimpleValueType(0);
+
+  MachineSDNode *N2 =
+      CurDAG->getMachineNode(NVPTX::I64toV2I32, SDLoc(N), EltVT, EltVT, Vector);
+
+  ReplaceNode(N, N2);
+  return true;
+}
+
```
Why can't this be done in `NVPTXInstrInfo.td`?

AlexMaclean replied:
I tried this but wasn't able to do it without llvm-tblgen crashing or hanging. I think table-gen support for nodes with multiple return values is not very reliable.

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2758,6 +2764,53 @@ static SDValue lowerCTLZCTPOP(SDValue Op, SelectionDAG &DAG) {
+  assert(A.getValueType() == MVT::i64 && B.getValueType() == MVT::i64);
+
+  const auto *AmtConst = dyn_cast<ConstantSDNode>(AmtVal);
+  if (!AmtConst)
+    return SDValue();
```
```suggestion
  if (!AmtConst) return SDValue();
```

AlexMaclean replied:
I don't think this is compliant with clang-format

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2758,6 +2764,53 @@ static SDValue lowerCTLZCTPOP(SDValue Op, SelectionDAG &DAG) {
+  const auto *AmtConst = dyn_cast<ConstantSDNode>(AmtVal);
+  if (!AmtConst)
+    return SDValue();
+  const auto Amt = AmtConst->getZExtValue() & 63;
```
> The shift argument is treated as an unsigned amount modulo the element size of the arguments. ([source](https://llvm.org/docs/LangRef.html#llvm-fshl-intrinsic))

Interesting semantics. I would've guessed it's a bug to specify a number larger than the element size.

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2758,6 +2764,53 @@ static SDValue lowerCTLZCTPOP(SDValue Op, SelectionDAG &DAG) {
   return DAG.getNode(ISD::ZERO_EXTEND, DL, MVT::i64, CT, SDNodeFlags::NonNeg);
 }
 
+static SDValue expandFSH64(SDValue A, SDValue B, SDValue AmtVal, SDLoc DL,
```
```suggestion
static SDValue expandFSH64(SDValue A, SDValue B, SDValue ShiftAmount, SDLoc DL,
```

AlexMaclean replied:
Done

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2758,6 +2764,53 @@ static SDValue lowerCTLZCTPOP(SDValue Op, SelectionDAG &DAG) {
+  SDValue BLo = UnpackB.getValue(0);
+  SDValue BHi = UnpackB.getValue(1);
+
+  // The bitfeild consists of { AHi : ALo : BHi : BLo }
+  // FSHL, Amt <  32 - The window will contain { AHi : ALo : BHi }
+  // FSHL, Amt >= 32 - The window will contain { ALo : BHi : BLo }
+  // FSHR, Amt <  32 - The window will contain { ALo : BHi : BLo }
+  // FSHR, Amt >= 32 - The window will contain { AHi : ALo : BHi }
```
Great comments :)

justinfargnoli reviewed (commented):
Can you add an additional section to your PR description where you describe why we need new `NVPTXISD` nodes to handle this case and under what conditions we can begin using `ISD` nodes and perhaps even port this to `LegalizeDAG`? 

Also, which test case is the regression?

AlexMaclean committed:
address comments

AlexMaclean commented:
> Can you add an additional section to your PR description where you describe why we need new `NVPTXISD` nodes to handle this case and under what conditions we can begin using `ISD` nodes and perhaps even port this to `LegalizeDAG`?

Added, also added a comment in the code.

> Also, which test case is the regression?

Every test case that changed show improvement. For example, in llvm/test/CodeGen/NVPTX/rotate_64.ll, `@rotate64` previously had 3 64-bit operations and now has 2 32-bit operations.

justinfargnoli commented:
> Every test case that changed show improvement.

Good point, I forgot that `mov.b64 {%r1, %r2}, %rd1;` is free.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2758,6 +2764,53 @@ static SDValue lowerCTLZCTPOP(SDValue Op, SelectionDAG &DAG) {
+  SDValue BLo = UnpackB.getValue(0);
+  SDValue BHi = UnpackB.getValue(1);
+
+  // The bitfeild consists of { AHi : ALo : BHi : BLo }
+  // FSHL, Amt <  32 - The window will contain { AHi : ALo : BHi }
+  // FSHL, Amt >= 32 - The window will contain { ALo : BHi : BLo }
+  // FSHR, Amt <  32 - The window will contain { ALo : BHi : BLo }
+  // FSHR, Amt >= 32 - The window will contain { AHi : ALo : BHi }
```
shifts of multiples of 32 bits are a special case here. We could directly build 64-bit result from 32-bit parts w/o further shifts. We could probably handle that in 32-bit shift, if we don't do that already. If we do, then it may be worth a comment that it's clean that not handling  whole word sized shifts here is deliberate. Based on the tests, it does look like whole-word shifts are handled via movs already, so we just want a comment to point to where that magic happens.

AlexMaclean replied:
Sounds good, I've added a comment clarifying that we rely on DAGCombiner to further optimize these cases for us:
```
  // Note that Amt = 0 and Amt = 32 are special cases where 32-bit funnel shifts
  // are not needed at all. Amt = 0 is a no-op producing either A or B depending
  // on the direction. Amt = 32 can be implemented by a packing and unpacking
  // move to select and arrange the 32bit values. For simplicity, these cases
  // are not handled here explicitly and instead we rely on DAGCombiner to
  // remove the no-op funnel shifts we insert.
```

Artem-B commented on llvm/test/CodeGen/NVPTX/rotate.ll:
```diff
@@ -265,32 +314,91 @@ define i64 @rotr64(i64 %a, i64 %n) {
+; SM20-EMPTY:
+; SM20-NEXT:  // %bb.0:
+; SM20-NEXT:    ld.param.u64 %rd1, [rotr64_32_imm_param_0];
+; SM20-NEXT:    shl.b64 %rd2, %rd1, 32;
+; SM20-NEXT:    shr.u64 %rd3, %rd1, 32;
+; SM20-NEXT:    or.b64 %rd4, %rd3, %rd2;
```
This is surprising. Why did we end up with the shifts and not moves? I guess is that we're falling through to the default implementation because funnel shifts are disabled on older GPUs, so we also miss the subset of cases that can be lowered as mov  on the older GPUs. 

We could mark funnel shifts as supported everywhere and bail out on non-multiples-of-32 shifts on older platforms.

AlexMaclean replied:
Yea, this could be a bit better optimized. That being said it would only benefit sm_20 and sm_32 both of which will hopefully be removed soon. As a result, I don't think it is worth the added complexity to handle this special case on such old platforms.

Artem-B reviewed (approved):
LGTM with a couple of nits.

AlexMaclean committed:
address comments
